src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```diff
@@ -0,0 +1,269 @@
+            throw new WebAuthnAuthenticationException("_", "User already exists");
+        }
+        final UserIdentity user = generateUserIdentity(username, displayName);
+        final StartRegistrationOptions startRegistrationOptions = StartRegistrationOptions.builder().user(user)
+                .authenticatorSelection(authenticatorSelection).timeout(TIMEOUT).build();
+        final PublicKeyCredentialCreationOptions options = rp.startRegistration(startRegistrationOptions);
+        final WebAuthnCredentialCreationInfo info = new WebAuthnCredentialCreationInfo();
+        info.setUsername(username);
+        info.setRealm(realm);
+        info.setOptions(options);
+        info.setProviderId(provider);
+        final String key = UUID.randomUUID().toString();
+        activeRegistrations.put(key, info);
+        final WebAuthnRegistrationResponse response = new WebAuthnRegistrationResponse();
+        response.setKey(key);
+        response.setOptions(options);
+        return response;
+    }
+
+    /**
+     * Returns:
+     * - the username of the authenticated user on successful authentication
+     * - null if the authentication was not successful
+     * - throws a WebAuthnAuthenticationException if some other error occourred
+     */
+    public String finishRegistration(
+            PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs> pkc,
+            String realm,
+            String key) throws WebAuthnAuthenticationException {
+        try {
+            final WebAuthnCredentialCreationInfo info = activeRegistrations.get(key);
+            WebAuthnAuthenticationException noSuchRegistration = new WebAuthnAuthenticationException("_",
+                    "Can not find matching active registration request");
+            if (info == null) {
+                throw noSuchRegistration;
+            }
+            if (!info.getRealm().equals(realm)) {
+                throw noSuchRegistration;
+            }
+            final String username = info.getUsername();
+            if (!StringUtils.hasText(username)) {
+                throw new WebAuthnAuthenticationException("_",
+                        "Could not finish registration: missing username");
+            }
+            WebAuthnUserAccount existingAccount = webAuthnUserAccountService.findByProviderAndUsername(provider,
+                    username);
+            if (existingAccount != null) {
+                throw new WebAuthnAuthenticationException("_",
+                        "Account already registered");
+            }
+            final PublicKeyCredentialCreationOptions options = info.getOptions();
+            RegistrationResult result = rp
+                    .finishRegistration(FinishRegistrationOptions.builder().request(options).response(pkc).build());
+            boolean attestationIsTrusted = result.isAttestationTrusted();
+            if (attestationIsTrusted) {
```

Comment:
if attestation is not trusted we should throw an exception

Reply:
Addressed in cb95a78885936ee33c8f8d0d9ad6d34aed10c7bf